File: app/src/main/java/com/gatheringhallstudios/mhworlddatabase/common/adapters/MonsterAdapterDelegate.java
package com.gatheringhallstudios.mhworlddatabase.common.adapters;

import android.support.annotation.NonNull;
import android.support.v7.widget.RecyclerView;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageView;
import android.widget.TextView;

import com.gatheringhallstudios.mhworlddatabase.R;
import com.gatheringhallstudios.mhworlddatabase.common.Consumer;
import com.gatheringhallstudios.mhworlddatabase.data.views.MonsterView;
import com.hannesdorfmann.adapterdelegates3.AdapterDelegate;

import java.util.List;

public class MonsterAdapterDelegate extends AdapterDelegate<List<MonsterView>> {

    Consumer<MonsterView> onSelected;

    public MonsterAdapterDelegate(Consumer<MonsterView> onSelected) {
        this.onSelected = onSelected;
    }

    @Override
    protected boolean isForViewType(@NonNull List<MonsterView> items, int position) {
        return true;
    }

    @NonNull
    @Override
    protected RecyclerView.ViewHolder onCreateViewHolder(ViewGroup parent) {
        LayoutInflater inflater = LayoutInflater.from(parent.getContext());
        View v = inflater.inflate(R.layout.listitem_monster, parent, false);

        return new MonsterViewHolder(v);
    }

    @Override
    protected void onBindViewHolder(@NonNull List<MonsterView> items,
                                    int position,
                                    @NonNull RecyclerView.ViewHolder holder,
                                    @NonNull List<Object> payloads) {
        MonsterView monster = items.get(position);

        MonsterViewHolder monVH = (MonsterViewHolder)holder;
        monVH.monsterName.setText(monster.name);

        holder.itemView.setOnClickListener((View v) -> {
            onSelected.accept(items.get(position));
        });

        // TODO Set monster image
    }

    class MonsterViewHolder extends RecyclerView.ViewHolder {
        ImageView monsterIcon;
        TextView monsterName;

        MonsterViewHolder(View itemView) {
            super(itemView);
            monsterIcon = itemView.findViewById(R.id.monster_icon);
            monsterName = itemView.findViewById(R.id.monster_name);
        }
    }
}

File: app/src/main/java/com/gatheringhallstudios/mhworlddatabase/common/adapters/MonsterAdapterDelegate.kt
package com.gatheringhallstudios.mhworlddatabase.common.adapters

import android.support.v7.widget.RecyclerView
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup

import com.gatheringhallstudios.mhworlddatabase.R
import com.gatheringhallstudios.mhworlddatabase.data.views.MonsterView
import com.gatheringhallstudios.mhworlddatabase.getAssetDrawable
import com.hannesdorfmann.adapterdelegates3.AdapterDelegate
import kotlinx.android.synthetic.main.listitem_monster.view.*

class MonsterAdapterDelegate(private val onSelected: (MonsterView) -> Unit) : AdapterDelegate<List<MonsterView>>() {

    override fun isForViewType(items: List<MonsterView>, position: Int): Boolean {
        return true
    }

    override fun onCreateViewHolder(parent: ViewGroup): RecyclerView.ViewHolder {
        val inflater = LayoutInflater.from(parent.context)
        val v = inflater.inflate(R.layout.listitem_monster, parent, false)

        return MonsterViewHolder(v)
    }

    override fun onBindViewHolder(items: List<MonsterView>,
                                  position: Int,
                                  holder: RecyclerView.ViewHolder,
                                  payloads: List<Any>) {
        val monster = items[position]

        val monVH = holder as MonsterViewHolder
        monVH.bind(monster)

        holder.itemView.setOnClickListener { onSelected(monster) }
    }

    internal inner class MonsterViewHolder(val view: View) : RecyclerView.ViewHolder(view) {

        fun bind(monster: MonsterView) {
            val defaultIcon = R.drawable.question_mark_grey
            val icon = view.context.getAssetDrawable(monster.data.icon, defaultIcon)
            view.monster_icon.setImageDrawable(icon)

            view.monster_name.text = monster.name
        }
    }
}
